scripts/datastore_upgrade.py

```
@@ -0,0 +1,475 @@
+STORE_DATASTORE_VERSION = 'Store Datastore version'
+CLOSE_ZKTRANSACTION = 'Close ZKTransaction connection'
+STOP_ZOO_KEEPER = 'Stop ZooKeeper'
+STOP_CASSANDRA = 'Stop Cassandra'
+VALIDATE_ENTITIES = 'Validate Entities'
+START_ZOOKEEPER = 'Start ZooKeeper'
+START_CASSANDRA = 'Start Cassandra'
+
+# The monit script to start the given service.
+START_SERVICE_SCRIPT = "python " + APPSCALE_HOME + "/scripts/monit_start_service.py "
+
+# The monit script to stop the given service.
+STOP_SERVICE_SCRIPT = "python " + APPSCALE_HOME + "/scripts/monit_stop_service.py "
+
+def ensure_app_is_not_running():
+  """ Ensures AppScale is not running as this is an offline script. """
+  logging.info("Ensure AppScale is not currently running...")
+  appscale_running = subprocess.call(['service', CONTROLLER_SERVICE, 'status']) == 0
+  if appscale_running:
+    logging.info("AppScale is running, please shut it down and try again.")
+    sys.exit(1)
+
+def start_cassandra(status_dict, db_ips, master_ip, keyname):
+  """ Creates a monit configuration file and prompts Monit to start Cassandra.
+  Args:
+    status_dict: A dictionary to record the status of the executed process.
+    db_ips: A list of database node IPs to start Cassandra on.
+    master_ip: The IP of the head node.
+    keyname: A string containing the deployment's keyname.
+  """
+  logging.info("Starting Cassandra...")
+  for ip in db_ips:
+    if ip == master_ip[0]:
+      cmd_status = monit_start_service.start_service(CASSANDRA_WATCH_NAME)
+    else:
+      start_service_cmd = START_SERVICE_SCRIPT + CASSANDRA_WATCH_NAME
+      cmd_status = utils.ssh(ip, keyname, start_service_cmd)
+
+    start_cassandra_ip = START_CASSANDRA + ":" + ip
+    if not cmd_status == 0:
+      logging.error("Monit was unable to start Cassandra.")
+      status_dict[start_cassandra_ip] = FAILURE
+      continue
+    logging.info("Successfully started Cassandra.")
+    status_dict[start_cassandra_ip] = SUCCESS
```

This will overwrite the failure status if a subsequent node is successful. The function should probably throw an exception if any node fails to start.